btf.c

```diff
@@ -192,8 +194,96 @@ btf_root_offset(struct btf *btf, const char *dotname)
 	return (btf_root_offset2(btf, dotname));
 }
 
+static struct quark_btf *
+quark_btf_new(const char *new_name)
+{
+	struct quark_btf	*qbtf;
+
+	if ((qbtf = malloc(sizeof(*qbtf) + sizeof(base_targets))) == NULL)
+		return (NULL);
+	qbtf->kname = strdup(new_name);
+	if (qbtf->kname == NULL) {
+		free(qbtf);
+		return (NULL);
+	}
+	memcpy(qbtf->targets, base_targets, sizeof(base_targets));
+
+	return (qbtf);
+}
+
+static struct quark_btf *
+quark_btf_dup(struct quark_btf *src)
+{
+	struct quark_btf	*qbtf;
+
+	if (src == NULL)
+		return (NULL);
+
+	qbtf = quark_btf_new(src->kname);
+	if (qbtf == NULL)
+		return (NULL);
+	memcpy(qbtf->targets, src->targets, sizeof(base_targets));
+
+	return (qbtf);
+}
+
 struct quark_btf *
-quark_btf_open(const char *path, const char *kname)
+quark_btf_open_hub(const char *version)
+{
+	extern struct quark_btf	 *all_btfs[];
+	struct quark_btf	**pp, *best, *cand;
+	int			  best_score;
+
+	/* paranoia */
+	if (version == NULL || strlen(version) == 0)
+		return (NULL);
+
+	best = NULL;
+	best_score = 0;
+	for (pp = all_btfs; (cand = *pp) != NULL; pp++) {
+		const char	*pv, *pc;
+		int		 score;
+
+		score = 0;
+
+		/* paranoia */
+		if (cand->kname == NULL || strlen(cand->kname) == 0)
+			return (NULL);
+
+		/*
+		 * Match head, the beginning of version
+		 */
+		for (pc = cand->kname, pv = version, score = 0;
```

Comment:
Multiple statements in a for expression is what makes all the fun in C, each for expression is in a different line.
I'd say it's relatively common C idiom, but yeah I can see how it can look a bit daunting :D

Reply:
I could break this into a big for loop with a bunch of conditionals, but it's more state imho.